### PR TITLE
fix(mme): Handling Maximum number of PDU sessions reached( Cause #65)

### PR DESCRIPTION
**Pull request by ganeshg87**
Test Cases:
    - verfied pdu session limit handling using ueransim
    - unit testcases added to validate functions

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp [hunk 1]
```diff
@@ -15,27 +15,34 @@
...
 #include "log.h"
 #include "conversions.h"
 #include "3gpp_38.401.h"
+#include "dynamic_memory_check.h"
 #ifdef __cplusplus
 }
 #endif
 #include "include/amf_session_manager_pco.h"
 #include "amf_recv.h"
+#include "./amf_sap.h"
```
Comment (lionelgo) [inline lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp]: file path here ?

**Comment by ganeshg87** [comment]
verified UT and UERANSIM.
please find logs and pcap attached 
[mme_max_pdu.log](https://github.com/magma/magma/files/7156123/mme_max_pdu.log)
image
image

File: lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GPayloadContainer.h [hunk 2]
```diff
@@ -12,13 +12,14 @@
 #pragma once
 #include <sstream>
 #include <cstdint>
+#include <string.h>
 #include <SmfMessage.h>
+#define PAYLOAD_CONTAINER_CONTENTS_MAX_LEN 8192
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GPayloadContainer.h:17]: are use using namespace std in this file? if so please remove `using namespace std` and explicitly use `std::`

**Review by ulaskozat — CHANGES_REQUESTED** [review]
please fix github action comments

File: lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GPayloadContainer.h (hunk 2 above)
Comment (ulaskozat) [reply lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GPayloadContainer.h:17]: if you are not using it at all, simply remove it.

Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GPayloadContainer.h:17]: the namespace is not using , removed it and updated review

Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GPayloadContainer.h:17]: @ulaskozat,  I have resolved the above review comment for complete nas5g module. but too many changes
                   can I include it in the same review?

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GPayloadContainer.cpp [hunk 3]
```diff
@@ -47,19 +47,25 @@ int PayloadContainerMsg::EncodePayloadContainerMsg(
...
+  }
+
+  IES_ENCODE_U16(buffer, tmp, encoded);
+
+  BUFFER_PRINT_LOG(payload_container->contents, encoded);
+  memcpy(buffer + tmp, payload_container->contents, encoded);
 
-  return (encoded);
+  return (encoded + tmp);
 };
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GPayloadContainer.cpp]: remove `;` after `}` as indicated by github actions.

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp [hunk 4]
```diff
@@ -517,6 +522,101 @@ TEST(test_amf_nas5g_pkt_process, test_amf_service_reject_message) {
...
+    /*
+     * Set header of plain NAS message
+     */
+    header->extended_protocol_discriminator = M5G_MOBILITY_MANAGEMENT_MESSAGES;
+    header->security_header_type = SECURITY_HEADER_TYPE_NOT_PROTECTED;
+  }
+
+  buffer = bfromcstralloc(len, "\0");
+  bytes  = nas5g_message_encode(buffer->data, &msg, len, nullptr);
+  EXPECT_TRUE(bytes > 0);
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf.cpp]: you should encode the message back and verify that it matches to the original message.

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GPayloadContainer.cpp (hunk 3 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GPayloadContainer.cpp]: addressed comment.

File: lte/gateway/c/core/oai/tasks/nas5g/include/SmfMessage.h [hunk 5]
```diff
@@ -29,6 +29,22 @@ class SmfMsgHeader {
...
   uint8_t message_type;
+
+  void copy(const SmfMsgHeader& s) {
+    extended_protocol_discriminator = s.extended_protocol_discriminator;
+    pdu_session_id                  = s.pdu_session_id;
+    procedure_transaction_id        = s.procedure_transaction_id;
+    message_type                    = s.message_type;
+  }
+  bool isEqual(const SmfMsgHeader& s) {
+    if ((extended_protocol_discriminator ==
```
Comment (lionelgo) [inline lte/gateway/c/core/oai/tasks/nas5g/include/SmfMessage.h]: Would suggest only bool isEqual(const SmfMsgHeader& s) const {

File: lte/gateway/c/core/oai/tasks/nas5g/include/SmfMessage.h [hunk 6]
```diff
@@ -57,5 +73,28 @@ class SmfMsg {
...
+    header.copy(s.header);
+    switch (s.header.message_type) {
+      case PDU_SESSION_ESTABLISHMENT_REQUEST:
+        msg.pdu_session_estab_request.copy(s.msg.pdu_session_estab_request);
+        break;
+      default:
+        break;
+    }
+  }
+  bool isEqual(const SmfMsg& s) {
```
Comment (lionelgo) [inline lte/gateway/c/core/oai/tasks/nas5g/include/SmfMessage.h:85]: Would suggest only bool isEqual(const SmfMsg& s) const {

File: lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GMessageType.h [hunk 7]
```diff
@@ -26,5 +26,13 @@ class MessageTypeMsg {
...
       MessageTypeMsg* message_type, uint8_t iei, uint8_t* buffer, uint32_t len);
+
+  void copy(const MessageTypeMsg& m) { msg_type = m.msg_type; }
+  bool isEqual(const MessageTypeMsg& m) {
+    if (msg_type == m.msg_type) {
+      return true;
+    }
+    return false;
+  }
 };
```
Comment (lionelgo) [inline lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GMessageType.h:32]: I suggest  return msg_type == m.msg_type;

Comment (lionelgo) [reply lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GMessageType.h:32]: and const method for all isEqual() methods

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp (hunk 4 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/test/amf/test_amf.cpp]: addressed review comment

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp (hunk 1 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp]: yes, headers and sources are on the same path.
but I have defined for one of warning, it looks odd, removed

File: lte/gateway/c/core/oai/tasks/nas5g/include/SmfMessage.h (hunk 5 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/nas5g/include/SmfMessage.h]: but isEqual api calling directly with instance variables of Each class. the "const member function" caller should be a constant variable. I am getting the below errors.
image

File: lte/gateway/c/core/oai/tasks/nas5g/include/SmfMessage.h (hunk 6 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/nas5g/include/SmfMessage.h:85]: comment fix not resolved

File: lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GMessageType.h (hunk 7 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GMessageType.h:32]: addressed the first comment.
2nd comment dropped

**Comment by ganeshg87** [comment]
@ulaskozat, @lionelgo,

     addressed review comments.Please provide your approval.

    -  max PDU limit verified with UERANSIM.
    -  updated AMF UT as Ulas suggested and verified.
[mme_max_pdu_21_sept.log](https://github.com/magma/magma/files/7198390/mme_max_pdu_21_sept.log)
image
image

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp [hunk 8]
```diff
@@ -24,6 +24,11 @@ extern "C" {
 #include "intertask_interface_types.h"
 #include "itti_free_defined_msg.h"
 }
+#if 0
+#define QUADLET 4
+#define AMF_GET_BYTE_ALIGNED_LENGTH(LENGTH)                                    \
+  LENGTH += QUADLET - (LENGTH % QUADLET)
+#endif
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf.cpp]: why do you have this?

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp [hunk 9]
```diff
@@ -517,6 +522,103 @@ TEST(test_amf_nas5g_pkt_process, test_amf_service_reject_message) {
...
+  len += dlmsg->payload_container.len;
+
+  /* Ciphering algorithms, EEA1 and EEA2 expects length to be mode of 4,
+   * so length is modified such that it will be mode of 4
+   */
+  AMF_GET_BYTE_ALIGNED_LENGTH(len);
+
+  buffer = bfromcstralloc(len, "\0");
+  bytes  = nas5g_message_encode(buffer->data, &msg, len, nullptr);
+  EXPECT_TRUE(bytes > 0);
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf.cpp]: legit github actions comment

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp (hunk 8 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/test/amf/test_amf.cpp]: addressed review comment

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp (hunk 9 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/test/amf/test_amf.cpp]: addressed review comment

**Review by ulaskozat — APPROVED** [review]
LGTM

**Pull request merged**